awx/main/managers.py

```diff
@@ -127,11 +129,21 @@ def register(self, uuid=None, hostname=None, ip_address=None, node_type='hybrid'
                         other_inst.save(update_fields=['ip_address'])
                         logger.warning("IP address {0} conflict detected, ip address unset for host {1}.".format(ip_address, other_hostname))
 
-            # Return existing instance that matches hostname
+            # get the instance based on the hostname
             instance = self.filter(hostname=hostname)
+
+            # Check or update the existing uuid
```

Comment:
It seems like the basic logic here is "default to looking up instance by uuid, if one is provided. otherwise, look up by hostname". Could make that more explicit with:

```
            # attempt to find instance by uuid
            if uuid is not None and uuid != UUID_DEFAULT:
                if self.filter(uuid=uuid).exists():
                    instance = self.filter(uuid=uuid)
            else:
                # otherwise, rely on hostname
                instance = self.filter(hostname=hostname)
```

Reply:
`update the existing uuid` doesn't seem accurate, since we're not updating the uuid here.

Reply:
what if there is a match based on both hostname & uuid.

In that case, you would prefer the uuid, but... you would change the hostname, and instances are unique on hostname. This would probably error.

This pattern (of creating fallback lookup fields in the midst of unique fields) and the errors that come with it are familiar to me from the inventory import `instance_id` field. If the precedence ever matters, then you're left with an IntegrityError.

Reply:
so currently this will change the hostname if it's different an a UUID and hostname are provided, I can see why that would be problematic, the solution would be to alert the user and verify that they want to change the hostname, but I can also see that getting very annoying very quickly.